### PR TITLE
Missions plugin improvement: Update missions list automatically 

### PR DESCRIPTION
**Pull request by NineBerry**
Missions plugin improvement: Update missions list automatically when moving the map

* Reuse one window for all mission lists
* Automatically update list when moving/zooming map
* Sort Missions by Name

### CODE REVIEW

File: plugins/missions.js [hunk 1]
```diff
@@ -190,17 +194,137 @@ window.plugin.missions = {
...
+						}
+					},
+					{
+						text: "Ok",
+						click: function() {
+							$(this).dialog('close');
+						}
+					},
+					{
+						text: "Missions in view",
```
Comment (modos189) [inline plugins/missions.js]: Can you explain what this button is for? I don't see any difference from removing it

File: plugins/missions.js [hunk 2]
```diff
@@ -219,7 +344,7 @@ window.plugin.missions = {
 		}).map(function(waypoint) {
 			return [waypoint.portal.latE6/1E6, waypoint.portal.lngE6/1E6];
 		});
-		
+
```
Comment (modos189) [inline plugins/missions.js]: Please remove deletion of blank lines. This and other lines below. If you want to reform the code (according to eslint rules) this should be done in a separate PR for all files

**Review by modos189 — CHANGES_REQUESTED** [review]

File: plugins/missions.js (hunk 2 above)
Comment (johnd0e) [reply plugins/missions.js]: > Please remove deletion of blank lines.

...and other white-space-only changes. At least in separate commit.
Otherwise you make reviewing meaningful changes harder.

Comment (NineBerry) [reply plugins/missions.js]: Whitespaces are back. Sorry for the multiple commits. I am not used to git. Still using subversion at work.

File: plugins/missions.js (hunk 1 above)
Comment (NineBerry) [reply plugins/missions.js]: The Missions list window can either show "missions in view" or missions starting at a specific portal.

When showing missions in view, the list is automatically updated when the map is scrolled or zoomed. While showing missions starting at a specific portal, the automatic update is disabled. This button is only shown when the list shows the missions starting at a specific portal and it will return the window to the mode where it shows the missions in view.

Comment (NineBerry) [reply plugins/missions.js]: Showing Missions in View

image

Showing Missions at one portal

image

File: plugins/missions.js [hunk 3]
```diff
@@ -238,7 +363,7 @@ window.plugin.missions = {
 				}
 				return;
 			}
-			callback(missions);
+			callback(missions, "Missions in View");
```
Comment (johnd0e) [inline plugins/missions.js]: Use single quotes https://github.com/IITC-CE/ingress-intel-total-conversion/wiki/Code-Style

File: plugins/missions.js [hunk 4]
```diff
@@ -904,6 +1045,24 @@ window.plugin.missions = {
...
 	},
 
+	onMoveEnd: function()
+	{
+		// When autorefresh is enabled (currently always)
+		// and not showing mission list of one portal
+		// and window not collpased
+		if(this.autoRefreshOnMoveEnd && !this.isShowingPortalList && !this.isMissionListCollapsed)
+		{
+			let dialog = window.DIALOGS['dialog-missionsList'];
```
Comment (johnd0e) [inline plugins/missions.js]: Use `var` https://github.com/IITC-CE/ingress-intel-total-conversion/wiki/Code-Style

And why not  `window.DIALOGS.dialog-missionsList`? Do we have some convention on such usage? (just a question, not request to change)

File: plugins/missions.js [hunk 5]
```diff
@@ -1026,6 +1185,8 @@ window.plugin.missions = {
 
 		window.addHook('search', this.onSearch.bind(this));
 
+		window.map.on('moveend', this.onMoveEnd.bind(this));
```
Comment (johnd0e) [inline plugins/missions.js]:
```suggestion
		window.map.on('moveend', this.onMoveEnd, this);
```

File: plugins/missions.js [hunk 6]
```diff
@@ -211,6 +335,7 @@ window.plugin.missions = {
 
 	zoomToMission: function(mission) {
 		map.fitBounds(this.getMissionBounds(mission), {maxZoom: DEFAULT_ZOOM});
+		// NOTE: Why maxZoom and not map.fitBounds(this.getMissionBounds(mission)); ???
```
Comment (johnd0e) [inline plugins/missions.js]: In order to prevent too close zoom

File: plugins/missions.js [hunk 7]
```diff
@@ -190,17 +194,137 @@ window.plugin.missions = {
...
+		{
+			$(plugin.missions.fromPortalListToNormalListButton).show();
+		}
+		else
+		{
+			$(plugin.missions.fromPortalListToNormalListButton).hide();
+		}
+
+		this.resizeMissionList();
+	},
```
Comment (johnd0e) [inline plugins/missions.js]: Use 'one true brace-style' https://eslint.org/docs/rules/brace-style#1tbs

File: plugins/missions.js (hunk 3 above)
Comment (NineBerry) [reply plugins/missions.js]: I made another commit to fix all the style guide issues

File: plugins/missions.js (hunk 4 above)
Comment (NineBerry) [reply plugins/missions.js]: Minus is not allowed in identifiers.

File: plugins/missions.js (hunk 5 above)
Comment (NineBerry) [reply plugins/missions.js]: Changed

File: plugins/missions.js (hunk 6 above)
Comment (NineBerry) [reply plugins/missions.js]: The zoom level is too far out here in my opinion. I have removed the comment and will create an issue for this later on

File: plugins/missions.js (hunk 7 above)
Comment (NineBerry) [reply plugins/missions.js]: I made another commit to fix all the style guide issues

File: plugins/missions.js [hunk 8]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+
+		// When showing missions in View and not portal mission list, refresh list now
+		if (!plugin.missions.isShowingPortalList) {
+			plugin.missions.openTopMissions();
+		}
+	},
+
+	resizeMissionList: function() {
+		if(!plugin.missions.isMissionListCollapsed)
+		{
```
Comment (johnd0e) [inline plugins/missions.js:291]: 1tbs

File: plugins/missions.js [hunk 9]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+				// A few lines down we will restrict height to fit the screen
+				$(openDialog).dialog({height: 'auto'});
+
+				// Restrict height of dialog to fit the screen
+				var dialogHeight = $(openDialog).parent().height();
+				var dialogTop = $(openDialog).parent().offset().top;
+				var mapHeight = $(window.map._container).height();
+
+				if(dialogTop + dialogHeight > mapHeight)
+				{
```
Comment (johnd0e) [inline plugins/missions.js:305]: 1tbs

File: plugins/missions.js [hunk 10]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+							$(this).dialog('close');
+						}
+					},
+					{
+						text: 'Missions in view',
+						css: {float: 'left', 'margin-right': '2px'},
+						click: function() {
+							plugin.missions.openTopMissions();
+						},
+						create: function()
```
Comment (johnd0e) [inline plugins/missions.js:245]: 1tbs

File: plugins/missions.js [hunk 11]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+				id: 'missionsList',
+				html: this.renderMissionList(missions),
+				height: 'auto',
+				width: '400px',
+				collapseCallback: this.onCollapseMissionList,
+				expandCallback: this.onExpandMissionList,
+				dragStop: this.resizeMissionList,
+				title: caption,
+				buttons:
+				[
```
Comment (johnd0e) [inline plugins/missions.js:226]:
```suggestion
				buttons:	[
```

File: plugins/missions.js [hunk 12]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+					{
+						text: 'Create new mission',
+						click: function() {
+							open('https://missions.ingress.com/');
+						}
+					},
+					{
```
Comment (johnd0e) [inline plugins/missions.js:233]: May be change such cases to `}, {`?

File: plugins/missions.js [hunk 13]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+
+			this.isMissionListCollapsed = false;
+		};
+
+		// Dialog will be open now
+		openDialog = window.DIALOGS['dialog-missionsList'];
+
+		// Set content and title
+		$(openDialog).html(this.renderMissionList(missions));
+		$(openDialog).dialog({title: caption});
```
Comment (johnd0e) [inline plugins/missions.js:262]:
```suggestion
		$(openDialog)
		  .html(this.renderMissionList(missions));
		  .dialog({title: caption});
```

**Review by johnd0e — CHANGES_REQUESTED** [review]

File: plugins/missions.js [hunk 14]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
-			'OK': function() { $(this).dialog('close'); },
-		});
+	showMissionListDialog: function(missions, caption, isPortalList) {
+
+		this.isShowingPortalList = isPortalList;
+
+		// Check whether dialog is already open
+		var openDialog = window.DIALOGS['dialog-missionsList'];
+		if (openDialog) {
+			// Dialog already there, expand if collpased
```
Comment (johnd0e) [inline plugins/missions.js:204]:
```suggestion
			// Dialog already there, expand if collapsed
```

File: plugins/missions.js [hunk 15]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+			// NOTE: It would be really cool if the IITC core had some function to read and modify collpased state of dialogs
+			// but it hasn't so we need to access internal information here
```
Comment (johnd0e) [inline plugins/missions.js:206]: Do we need this comment here?
And are you sure that we can't  access internal info?

File: plugins/missions.js [hunk 16]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+			if (plugin.missions.isMissionListCollapsed) {
+				var dia   = $(openDialog).closest('.ui-dialog');
+				var button   = dia.find('.ui-dialog-titlebar-button-collapse');
```
Comment (johnd0e) [inline plugins/missions.js:209]: Redundant spaces

File: plugins/missions.js [hunk 17]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+		// Set content and title
+		$(openDialog).html(this.renderMissionList(missions));
+		$(openDialog).dialog({title: caption});
+
+		// When showing list for one portal, show button to switch back to general list, otherwise hide it
+		if (isPortalList) {
+			$(plugin.missions.fromPortalListToNormalListButton).show();
+		} else {
+			$(plugin.missions.fromPortalListToNormalListButton).hide();
+		}
```
Comment (johnd0e) [inline plugins/missions.js:269]: Matter of taste, but I'd prefer that shorter
`	 $(plugin.missions.fromPortalListToNormalListButton)[isPortalList ? 'show' : 'hide']();`

File: plugins/missions.js [hunk 18]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+	},
+
+	resizeMissionList: function() {
+		if(!plugin.missions.isMissionListCollapsed)
+		{
+			var openDialog = window.DIALOGS['dialog-missionsList'];
+
+			if (openDialog) {
+				// Make dialog choose height automatically based on content first
+				// A few lines down we will restrict height to fit the screen
```
Comment (johnd0e) [inline plugins/missions.js:296]: Not sure that we need special code to handle this here.

File: plugins/missions.js [hunk 19]
```diff
@@ -190,17 +194,124 @@ window.plugin.missions = {
...
+				$(openDialog).dialog({height: 'auto'});
+
+				// Restrict height of dialog to fit the screen
+				var dialogHeight = $(openDialog).parent().height();
+				var dialogTop = $(openDialog).parent().offset().top;
+				var mapHeight = $(window.map._container).height();
```
Comment (johnd0e) [inline plugins/missions.js:302]: https://leafletjs.com/reference-1.7.1.html#map-getsize

File: plugins/missions.js [hunk 20]
```diff
@@ -904,6 +1031,21 @@ window.plugin.missions = {
 		}
 	},
 
+	onMoveEnd: function() {
+		// When autorefresh is enabled (currently always)
+		// and not showing mission list of one portal
+		// and window not collpased
```
Comment (johnd0e) [inline plugins/missions.js:1037]:
```suggestion
		// and window not collapsed
```

File: plugins/missions.js [hunk 21]
```diff
@@ -904,6 +1031,21 @@ window.plugin.missions = {
...
+		if (this.autoRefreshOnMoveEnd && !this.isShowingPortalList && !this.isMissionListCollapsed) {
+			var dialog = window.DIALOGS['dialog-missionsList'];
+
+			// and if dialog is visible
+			if (dialog) {
```
Comment (johnd0e) [inline plugins/missions.js:1042]:
```suggestion
			if (dialog) { // is visible
```

File: plugins/missions.js [hunk 22]
```diff
@@ -904,6 +1031,21 @@ window.plugin.missions = {
...
+				// then refresh the mission list
+				this.openTopMissions();
```
Comment (johnd0e) [inline plugins/missions.js:1044]:
```suggestion
				this.openTopMissions(); // refresh 
```

**Review by johnd0e — CHANGES_REQUESTED** [review]

**Comment by modos189** [comment]
@NineBerry Hi, what do you think of johnd0e suggestions?

**Comment by NineBerry** [comment]
> @NineBerry Hi, what do you think of johnd0e suggestions?

It's mostly code formatting issues. There were other commits to the file by someone else that changed the code formatting. I'll adapt my changes accordingly as sonn as I find the time to do so.

With regard to the issue of automatically limiting the size of the "missions in view" dialog. It's necessary because using the current situation, the dialog changes size vertically automatically which means that when the list is refreshed automatically, the only way to be able to see all missions is by having the dialog at the top of the screen.
